cmd/restic/cmd_serve.go

```diff
@@ -0,0 +1,268 @@
+				}
+				return
+			}
+		}
+
+		if len(files) == 1 && files[0].Node.Type == "file" {
+			if err := dump.New("zip", repo, w).WriteNode(ctx, files[0].Node); err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			return
+		}
+
+		var rows []treePageRow
+		for _, item := range files {
+			if item.Path != curPath {
+				rows = append(rows, treePageRow{"/tree/" + snapshotID + item.Path, item.Node.Name, item.Node.Type, item.Node.Size, item.Node.ModTime})
+			}
+		}
+		sort.SliceStable(rows, func(i, j int) bool {
+			return strings.ToLower(rows[i].Name) < strings.ToLower(rows[j].Name)
+		})
+		sort.SliceStable(rows, func(i, j int) bool {
+			return rows[i].Type == "dir" && rows[j].Type != "dir"
+		})
+		parent := "/tree/" + snapshotID + curPath + "/.."
+		if curPath == "/" {
+			parent = "/"
+		}
+		if err := treePage.Execute(w, treePageData{snapshotID + ": " + curPath, parent, rows}); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	})
+
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/" {
+			http.NotFound(w, r)
+			return
+		}
+		var rows []indexPageRow
+		for sn := range FindFilteredSnapshots(ctx, repo.Backend(), repo, &restic.SnapshotFilter{}, nil) {
+			rows = append(rows, indexPageRow{"/tree/" + sn.ID().Str() + "/", sn.ID().Str(), sn.Time, sn.Hostname, sn.Tags, sn.Paths})
+		}
+		sort.Slice(rows, func(i, j int) bool {
+			return rows[i].Time.After(rows[j].Time)
+		})
+		if err := indexPage.Execute(w, indexPageData{"Snapshots", rows}); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	})
+
+	http.HandleFunc("/style.css", func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Cache-Control", "max-age=300")
+		_, _ = w.Write([]byte(stylesheetTxt))
+	})
+
+	Printf("Now serving the repository at http://%s\n", opts.Listen)
+	Printf("When finished, quit with Ctrl-c here.\n")
+
+	return http.ListenAndServe(opts.Listen, nil)
+}
+
+type indexPageRow struct {
+	Link  string
+	ID    string
+	Time  time.Time
+	Host  string
+	Tags  []string
+	Paths []string
+}
+
+type indexPageData struct {
+	Title string
+	Rows  []indexPageRow
+}
+
+type treePageRow struct {
+	Link string
+	Name string
+	Type string
+	Size uint64
+	Time time.Time
+}
+
+type treePageData struct {
+	Title  string
+	Parent string
+	Rows   []treePageRow
+}
+
+const indexPageTpl = `<html>
+<head>
+<link rel="stylesheet" href="/style.css">
+<title>{{.Title}} :: restic</title>
+</head>
+<body>
+<h1>{{.Title}}</h1>
+<table>
+<thead><tr><th>ID</th><th>Time</th><th>Host</th><th>Tags</th><th>Paths</th></tr></thead>
+<tbody>
+{{range .Rows}}
+<tr><td><a href="{{.Link}}">{{.ID}}</a></td><td>{{.Time | FormatTime}}</td><td>{{.Host}}</td><td>{{.Tags}}</td><td>{{.Paths}}</td></tr>
+{{end}}
+</tbody>
+</table>
+</body>
+</html>`
+
+const treePageTpl = `<html>
+<head>
+<link rel="stylesheet" href="/style.css">
+<title>{{.Title}} :: restic</title>
```

Comment:
Uhm, what about adding viewport meta tags, so that mobile browsers are happy?
https://developer.mozilla.org/en-US/docs/Web/HTML/Viewport_meta_tag

```html
<meta name="viewport" content="width=device-width, initial-scale=1" />
```

I haven't tested the current code, though.